### DISCUSSION TITLE
How to clear AutoCompleteTextBox

### DISCUSSION THREAD

**Discussion by Enat0** [HandyOrg/HandyControl#1279] [Q&A]
Hello,

How can I clear the text of AutoCompleteTextBox with MVVM?
I'm trying to clear a form if I press a button but I can't seem to empty the text.. 

Here's my code:

```
<hc:AutoCompleteTextBox 
Width="210" Height="30"  
FontFamily="Calibri" FontSize="13" 
Foreground="{DynamicResource BorderBrushGeral} 
BorderBrush="{DynamicResource BorderBrushGeral}"
ItemsSource="{Binding ListaTiposSistema}" 
SelectedValue="{Binding PesquisaDefinida.Categoria}"
SelectedItem="{Binding PesquisaDefinida.Categoria}"
Text="{Binding NomeTiposSistema}"
DisplayMemberPath="NomesTiposSistema" SelectedValuePath="NomesTiposSistema"
 MaxDropDownHeight="339"/>
```

And in my model view:

```
// ------------------------ Tipos Sistemas --------------- INICIO
public ManualObservableCollection<TiposSistemaPesquisa> ListaTiposSistema { get; set; }
    private List<TiposSistemaPesquisa> _tiposSistemaLista;
        public class TiposSistemaPesquisa
        {
            public string NomesTiposSistema { get; set; }
        }

        private string _nomeTiposSistema;
        public string NomeTiposSistema
        {
            get => _nomeTiposSistema;
            set
            {
                SetField(ref _nomeTiposSistema, value);
                PesquisaDefinida.Categoria = value;
                FiltrarTiposSistema(value);
            }
        }

        private void FiltrarTiposSistema(string key)
        {
            ListaTiposSistema.CanNotify = false;
            ListaTiposSistema.Clear();
            if (_tiposSistemaLista != null)
            {
                foreach (TiposSistemaPesquisa data in _tiposSistemaLista)
                {
                    if (data.NomesTiposSistema.ToLower().Contains(key.ToLower()))
                    {
                        ListaTiposSistema.Add(data);
                    }
                }
            }
            ListaTiposSistema.CanNotify = true;
        }

        private async void GetTipoSistemasAsync()
        {
            _tiposSistemaLista = await Task.Run(() => TipoSistemaData.GetTipoSistemaPesquisa());
            if (PesquisaDefinida.Categoria != null)
            {
                ListaTiposSistema.Add(new TiposSistemaPesquisa { NomesTiposSistema = PesquisaDefinida.Categoria });
                _nomeTiposSistema = PesquisaDefinida.Categoria;
            }
        }
```

**Comment by JeremyWu917** [comment]
@Enat0 
Hi there,
You can find the way to clear AutoCompleteTextBox in `HandyControl` demo.
Here is the key:
```C#
using System.Collections.Generic;
using GalaSoft.MvvmLight;
using HandyControl.Collections;
using HandyControlDemo.Data;
using HandyControlDemo.Service;

namespace HandyControlDemo.ViewModel;

public class AutoCompleteTextBoxDemoViewModel : ViewModelBase
{
    private string _searchText;

    public string SearchText
    {
        get => _searchText;
#if NET40
        set
        {
            Set(nameof(SearchText), ref _searchText, value);
            FilterItems(value);
        }
#else
        set
        {
            Set(ref _searchText, value);
            FilterItems(value);
        }
#endif
    }

    public ManualObservableCollection<DemoDataModel> Items { get; set; } = new();

    private readonly List<DemoDataModel> _dataList;

    public AutoCompleteTextBoxDemoViewModel(DataService dataService)
    {
        _dataList = dataService.GetDemoDataList(10);
    }

    private void FilterItems(string key)
    {
        Items.CanNotify = false;

        Items.Clear();

        foreach (var data in _dataList)
        {
            if (data.Name.ToLower().Contains(key.ToLower()))
            {
                Items.Add(data);
            }
        }

        Items.CanNotify = true;
    }
}

```

In your case, you just need to diynamic clear `NomeTiposSistema` or `ListaTiposSistema`, _but_, you should noted if you clear `ListaTiposSistema` means you should to refill data to `ListaTiposSistema` in next step.

Hope it works.
Thanks

**Reply by Enat0** [reply]
Thanks for the reply but that didn't work.
I made a gif trying to explain.. Everything clears except the autotextbox:

[http://g.recordit.co/LekUT6S5Tx.gif](url)

**Reply by JeremyWu917** [reply]
@Enat0 
Hi
`PesquisaDefinida.Categoria` sholud be cleared too and the property of `PesquisaDefinida.Categoria` sholud be MVVM two-way.

Thanks

**Reply by Enat0** [reply]
In the GIF I posted I didn't have anything in "Categoria" field, but "Posto" uses AutocompleteTextBox too and it isn't working the way you said.

The button "Limpar Critérios da Pesquisa" is a clear button and in MVVM I have this:

```
public PesquisaPostos LimparPesquisa()
        {
            try
            {
                string utilizador = PesquisaDefinida.Utilizador;
                PesquisaDefinida = new PesquisaPostos
                {
                    DaData = DateTime.Now,
                    AData = DateTime.Now,
                    Utilizador = utilizador,
                    Posto = "",
                    Categoria = ""
                };
                NomePosto = "";
                PesquisaDefinida.Posto = "";
                NomeTiposSistema = "";
                ListaTiposSistema.Clear();
            }
            catch(Exception ex)
            {
                _ = MessageBox.Show(ex.Message);
            }
            return PesquisaDefinida;
        }
```

I've tried create a erase button in the demo but it isn't erasing the text in the textbox

Edit: maybe I'm explaining it wrong, English is not my native language.. I don't want to clear the suggestion box that appears, I only want to erase the text the user type

**Reply by JeremyWu917** [reply]
> In the GIF I posted I didn't have anything in "Categoria" field, but "Posto" uses AutocompleteTextBox too and it isn't working the way you said.
> 
> The button "Limpar Critérios da Pesquisa" is a clear button and in MVVM I have this:
> 
> ```
> public PesquisaPostos LimparPesquisa()
>         {
>             try
>             {
>                 string utilizador = PesquisaDefinida.Utilizador;
>                 PesquisaDefinida = new PesquisaPostos
>                 {
>                     DaData = DateTime.Now,
>                     AData = DateTime.Now,
>                     Utilizador = utilizador,
>                     Posto = "",
>                     Categoria = ""
>                 };
>                 NomePosto = "";
>                 PesquisaDefinida.Posto = "";
>                 NomeTiposSistema = "";
>                 ListaTiposSistema.Clear();
>             }
>             catch(Exception ex)
>             {
>                 _ = MessageBox.Show(ex.Message);
>             }
>             return PesquisaDefinida;
>         }
> ```
> 
> I've tried create a erase button in the demo but it isn't erasing the text in the textbox
> 
> Edit: maybe I'm explaining it wrong, English is not my native language.. I don't want to clear the suggestion box that appears, I only want to erase the text the user type

All right, I'll post a demo for you to ref ☕

**Comment by JeremyWu917** [comment]
> NomeTiposSistema

Sorry for the later response.

This part of ViewModel should inherit `ViewModelBase` or other class that inherit `INotifyPropertyChanged`.
This ensures that `NomeTiposSistema` is a two-way notification property.

From now on I guess is that it should not have been implemented.

Here is HandyControl's Demo:

```C#
public class AutoCompleteTextBoxDemoViewModel : ViewModelBase
{

}

//
// Summary:
//     A base class for the ViewModel classes in the MVVM pattern.
public abstract class ViewModelBase : ObservableObject, ICleanup
{

}

//
// Summary:
//     A base class for objects of which the properties must be observable.
public class ObservableObject : INotifyPropertyChanged
{

}
```

So, could you please post a demo about you issues.

Thanks.

**Reply by Enat0** [reply]
Thanks again for the reply and sorry for all the trouble..

My class is inheriting ViewModelBase and it's inheriting INotifyPropertyChanged.

I have a demo that I uploaded to mediafire (If it's against the rules tell me where I can upload it please). [Here's the link](https://www.mediafire.com/folder/dfw9cgfhtiz6g/AutoCompleteTextBox)

I have one autocompletetextbox, one textbox and one button. The button should clear both AutoCompletetextbox and textbox, but its only clearing textbox.

One thing I noticed is I can clear AutoCompleteTextBox if I change the Style.. Maybe there's something wrong with the style?

**Reply by JeremyWu917** [reply]
> Thanks again for the reply and sorry for all the trouble..
> 
> My class is inheriting ViewModelBase and it's inheriting INotifyPropertyChanged.
> 
> I have a demo that I uploaded to mediafire (If it's against the rules tell me where I can upload it please). [Here's the link](https://www.mediafire.com/folder/dfw9cgfhtiz6g/AutoCompleteTextBox)
> 
> I have one autocompletetextbox, one textbox and one button. The button should clear both AutoCompletetextbox and textbox, but its only clearing textbox.
> 
> One thing I noticed is I can clear AutoCompleteTextBox if I change the Style.. Maybe there's something wrong with the style?

Copy, thanks.

And i got the same issues 🤣 

@ghost1372 
Hi bro, could you please have a look.

AutoCompleteTextBox's property: Text can not notify from vm to view. [here](https://github.com/ghost1372/HandyControls/blob/9b441090bbdd9303b630d0721156e62210c97d78/src/Shared/HandyControl_Shared/Controls/Input/AutoCompleteTextBox/AutoCompleteTextBox.cs#L22) maybe the issues fire.

Thanks.

**Reply by JeremyWu917** [reply]
> > Thanks again for the reply and sorry for all the trouble..
> > My class is inheriting ViewModelBase and it's inheriting INotifyPropertyChanged.
> > I have a demo that I uploaded to mediafire (If it's against the rules tell me where I can upload it please). [Here's the link](https://www.mediafire.com/folder/dfw9cgfhtiz6g/AutoCompleteTextBox)
> > I have one autocompletetextbox, one textbox and one button. The button should clear both AutoCompletetextbox and textbox, but its only clearing textbox.
> > One thing I noticed is I can clear AutoCompleteTextBox if I change the Style.. Maybe there's something wrong with the style?
> 
> Copy, thanks.
> 
> And i got the same issues 🤣
> 
> @ghost1372 Hi bro, could you please have a look.
> 
> AutoCompleteTextBox's property: Text can not notify from vm to view. [here](https://github.com/ghost1372/HandyControls/blob/9b441090bbdd9303b630d0721156e62210c97d78/src/Shared/HandyControl_Shared/Controls/Input/AutoCompleteTextBox/AutoCompleteTextBox.cs#L22) maybe the issues fire.
> 
> Thanks.

@ghost1372 Any update? thanks

**Reply by ghost1372** [reply]
Hi, I am very busy these days I will try to take a look as soon as possible
Tnx
